src/SAWScript/Builtins.hs

```diff
@@ -1226,6 +1227,151 @@ provePrim script t = do
+      case ttType relation of
+        TypedTermSchema
+          (C.Forall
+            []
+            []
+            (C.TCon
+              (C.TC C.TCFun)
+              [ C.TCon (C.TC (C.TCTuple 2)) [s1, o1]
+              , C.TCon
+                (C.TC C.TCFun)
+                [ C.TCon (C.TC (C.TCTuple 2)) [s2, o2]
+                , C.TCon (C.TC C.TCBit) []]])) -> do
+          unless (o1 == o2) $ fail $ unlines
+            [ "Error: Mismatched output types in relation."
+            , "LHS output type: " ++ Cryptol.pretty o1
+            , "RHS output type: " ++ Cryptol.pretty o2 ]
+
+          return (s1, s2, o1)
+        _ -> fail $ "Error: Unexpected relation type: "
+                 ++ show (ppTypedTermType (ttType relation))
+
+    -- Typecheck bisimulation term. Returns the term's output type.
+    typecheckSide :: TypedTerm -> C.Type -> C.Type -> TopLevel C.Type
+    typecheckSide side stateType outputType =
+      case ttType side of
+        TypedTermSchema
+          (C.Forall
+            []
+            []
+            (C.TCon
+              (C.TC C.TCFun)
+              [ C.TCon (C.TC (C.TCTuple 2)) [s, i]
+              , C.TCon (C.TC (C.TCTuple 2)) [s', o] ])) -> do
+          unless (s == stateType) $ fail $ unlines
+            [ "Error: State type in bisimulation term input does not match state type in relation."
+            , "  Expected: " ++ Cryptol.pretty stateType
+            , "  Actual: " ++ Cryptol.pretty s]
+
+          unless (s' == stateType) $ fail $ unlines
+            [ "Error: State type in bisimulation term output does not match state type in relation."
+            , "  Expected: " ++ Cryptol.pretty stateType
+            , "  Actual: " ++ Cryptol.pretty s']
+
+          unless (o == outputType) $ fail $ unlines
+            [ "Error: Output type in bisimulation term does not match output type in relation."
+            , "  Expected: " ++ Cryptol.pretty outputType
+            , "  Actual: " ++ Cryptol.pretty o ]
+
+          return i
+        _ ->
+          let stStr = Cryptol.pretty stateType in
+          fail $ unlines
+          [ "Error: Unexpected bisimulation term type."
+          , "  Expected: (" ++ stStr ++ ", inputType) -> (" ++ stStr ++ ", outputType)"
+          , "  Actual: " ++ show (ppTypedTermType (ttType side)) ]
+
+    -- Generate 'Term' for application of a relation
+    scRelation :: SharedContext -> Term -> Term -> TopLevel Term
+    scRelation sc relLhs relRhs = io $
+      scApplyAll sc (ttTerm relation) [relLhs, relRhs]
+
+    -- Given a 'LocalName' and 'Type' for a function argument, return the same
+    -- 'LocalName' and a 'Term' for the argument type. This function is intended
+    -- to be used in conjunction with 'scLambdaList'.
+    importArg :: SharedContext
+              -> (LocalName, C.Type)
+              -> TopLevel (LocalName, Term)
+    importArg sc (name, t) = do
+      t' <- io $ Cryptol.importType sc Cryptol.emptyEnv t
+      return (name, t')
```

Comment:
I don't think the `name` part of the tuple is used anywhere in the function. Would it be cleaner to remove it and just have the type `SharedContext -> C.Type -> TopLevel Term`?

Reply:
I made this function to try to simplify a `mapM` that needs to preserve the first element in a tuple and modify the second, but I think it wound up being more confusing to follow what's going on. I removed this and just inlined it.